add tables of marks

ThomasBreuer opened the pull request:
(I need this for the forthcoming talk at ICMS.)

- code
- tests
- documentation
- and improve the caching of character tables from the library: normalize the key such that the Oscar table gets created just once from the underlying GAP table

ThomasBreuer committed:
add tables of marks

(I need this for the forthcoming talk at ICMS.)

- code
- tests
- documentation

ThomasBreuer committed:
fixed a doctest

and added a few more `GAP.Globals` -> `GAPWrap`

ThomasBreuer committed:
add tests (increase code coverage)

fingolfin commented on docs/oscar_references.bib:
```diff
@@ -386,6 +386,16 @@ @Article{Bis96
+  author        = {Burnside, W.},
+  title         = {Theory of groups of finite order},
+  mrnumber      = {69818},
+  note          = {2d ed},
```
Just a guess, but maybe it should be this?
```suggestion
  note          = {2nd ed},
```

ThomasBreuer replied:
If depends.
If we want to be consistent with [MathSciNet](https://mathscinet.ams.org/mathscinet/article?mr=69818) then we should keep it as it is.

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+  isomorphism::Map  # isomorphism from `group` to a group in GAP
+
+  function GAPGroupTableOfMarks(G::Union{GAPGroup, FinGenAbGroup}, tom::GapObj, iso::Map)
+    return new(tom, G, iso)
```
Maybe add here and in the next constructor a check verifying that `tom` actually is a table of marks?

ThomasBreuer replied:
O.k.

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+    gaptom = GAPWrap.TableOfMarks(codomain(iso))
+    return GAPGroupTableOfMarks(G, gaptom, iso)
+  end
+end
```
To make this type stable and also a bit more concise, how about using the `@attr` property:
```suggestion
@attr GAPGroupTableOfMarks function table_of_marks(G::Union{GAPGroup, FinGenAbGroup})
  iso = isomorphism_to_GAP_group(G)
  gaptom = GAPWrap.TableOfMarks(codomain(iso))
  return GAPGroupTableOfMarks(G, gaptom, iso)
end
```

ThomasBreuer replied:
O.k. (as a continuation of #3927)

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+# Tables of marks from the library are cached in the dictionary
+# `tables_of_marks_by_id`,
+# in order to achieve that fetching the same table twice yields the same
+# object.
```
Then perhaps use a `AbstractAlgebra.WeakValueCache` (resp. its alias `AbstractAlgebra.CacheDictType`)

ThomasBreuer replied:
Then it may happen that the Julia objects get garbage collected if no variable keeps them alive, and the underlying GAP objects are still cached on the GAP side, thus creating the Julia object anew will yield different Julia objects pointing to identical GAP objects. I do not see a problem with that, but it looks strange.

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+    table_of_marks(id::String)
+
+Return the table of marks
+for which `id` is an admissible name in GAP's library of tables of marks.
```
Is there a cheap way to test if a string is an admissible name?

Is there a way to get a list of all admissible names?

fingolfin reviewed (approved):
oops I just noticed I never sent the comments I added during the weekend :-(.

Anyway, these are all minor things and could also be addressed in a follow-up PR, if at all

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+      return GAPGroupTableOfMarks(G, tom, iso)
+    else
+      return GAPGroupTableOfMarks(tom)
+    end
```
```suggestion
    end::GAPGroupTableOfMarks
```
to make it type stable / allow Julia to determine the return type correctly

ThomasBreuer replied:
o.k.

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+      pos !== GAP.Globals.fail && return character_table(string(info[2][pos]))
+    end
+    return character_table(group(tom))
+  end
```
```suggestion
  end::GAPGroupCharacterTable
```

ThomasBreuer replied:
o.k.

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+    return character_table(group(tom))
+  end
+
+  set_attribute!(tbl, :table_of_marks, tom)
```
Can't you move that into the do-end block? Then this method could also make use of `@attr`

ThomasBreuer replied:
yes, this works

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+  tom = get_attribute!(tbl, :table_of_marks) do
+    isdefined(tbl, :group) && return table_of_marks(group(tbl))
+    return table_of_marks(identifier(tbl))
+  end
```
```suggestion
  end::GAPGroupTableOfMarks
```

fingolfin commented on src/Groups/tom.jl:
```diff
@@ -0,0 +1,613 @@
+    return table_of_marks(identifier(tbl))
+  end
+
+  set_attribute!(tom, :character_table, tbl)
```
As above, if this was moved into the do-end block, this method could use `@attr`

ThomasBreuer committed:
address comments

- provide `all_table_of_marks_names` for computing admissible names
  of tables of marks from the library,
  analogous to (and based on) `all_character_table_names`;
  this is something which GAP does not offer

- improve the documentation of admissible names for character tables,
  refer to this documentation when talking about admissible names for
  tables of marks

- add `is_table_of_marks_name` and `is_character_table_name`

- change the setup for names of tables of marks from the library:
  Use always the name of the corresponding library character table.
  For that, turn `identifier` for tables of marks into an attribute,
  and provide auxiliary functions for translating between the `Identifier`
  values for character tables and tables of marks on the GAP side.

ThomasBreuer committed:
fix a docstring
